src/Language/PureScript/Types.hs

```diff
@@ -516,39 +516,48 @@ isMonoType _        = True
 mkForAll :: [(a, (Text, Maybe (Type a)))] -> Type a -> Type a
 mkForAll args ty = foldr (\(ann, (arg, mbK)) t -> ForAll ann TypeVarInvisible arg mbK t Nothing) ty args
 
+-- | Rename a type variable, taking into account variable shadowing
+renameTypeVar :: Text -> Text -> Type a -> Type a
+renameTypeVar original replacement = replaceAllTypeVars' [(original, Left replacement)]
+
 -- | Replace a type variable, taking into account variable shadowing
 replaceTypeVars :: Text -> Type a -> Type a -> Type a
 replaceTypeVars v r = replaceAllTypeVars [(v, r)]
 
 -- | Replace named type variables with types
 replaceAllTypeVars :: [(Text, Type a)] -> Type a -> Type a
-replaceAllTypeVars = go [] where
-  go :: [Text] -> [(Text, Type a)] -> Type a -> Type a
-  go _  m (TypeVar ann v) = fromMaybe (TypeVar ann v) (v `lookup` m)
+replaceAllTypeVars rs = replaceAllTypeVars' (fmap (fmap Right) rs)
+
+replaceAllTypeVars' :: [(Text, Either Text (Type a))] -> Type a -> Type a
```

Comment:
It seems like the reason you made an `Either` version of this function is that you want to preserve the annotation on the original type var when you rename it? I'm not convinced that's desirable; the new name we're introducing is synthetic and the annotation should reflect that. What's the argument for?

Reply:
My goal here was to simply reuse most of the original implementation of `replaceAllTypeVars` to implement `renameTypeVar`. In that context, I don't have a `Type a`, only a `Text`. However, the original `replaceAllTypeVars` assumes I have a `Type a`. So, I updated what I could and left everything else unchanged. That required wrapping the arg in an `Either` to support both use cases. If the original source annotation is unnecessary, I didn't know that.

Reply:
If there isn't a need to copy the source annotation from the natural variable to the synthetic one, isn't it a lot simpler just to create a `TypeVar` and reuse `replaceAllTypeVars` as-is?

I guess since these are annotation-polymorphic functions, `moveQuantifiersToFront` would have to accept an `a` to use, but that seems less disruptive than embedding an `Either` in `replaceAllTypeVars` and giving it two interfaces.

Reply:
Fair enough.